feat(api): gripper error detection

sfoster1 opened the pull request:
# Description
Overhauls gripper error detection to take into account various sources of error.

There are three sources of error in determining whether the gripper jaws are at a location that means it probably did (or didn't) successfully grip something:

## Where are the gripper jaws?

Grippers are manufactured with a critical dimension hard stop - when the jaws are closed. When the jaws are opened, the jaws may be in different places on different grippers. We say that the jaw is "homed" - and zero its encoder - when the jaw is opened, and then calculate position by subtracting the current encoder value from the "known" open position. That means that you might close a gripper all the way to the dimensionally accurate stop and see that it thinks it is at 58mm or 62mm instead of 60mm, because that unit's open position was actually physically 90mm or 94mm instead of 92mm.

The way to fix it is to run a quick calibration - open the jaws to the stops set the encoder zero; close the jaws to the stops and capture the position; save the difference between the captured position and the nominal jaw displacement as an offset. It's really fast, and since it's really fast and dependent on the gripper we can just do it whenever we home the gripper for the first time. Then we'll keep it in memory and use it for positioning the jaws, and implicitly fold it in to position calculations including critical point calculations for things on the jaws like the calibration probe.

Note that neither this change to add gripper jaw calibration nor the previous incorrect lack of gripper jaw calibration affects gripper _offset_ calibration, because the two jaws are equally and oppositely affected by the manufacturing variance, and their position is averaged.

fa1e3511e1e92505f22e3dadf0cd219686ec2fed is this change.

## How wide is the thing where the gripper gripped it?

Our labware definitions capture a bounding box for the labware as a whole - a rectangular prism that is as small as it can be while still encompassing the entire labware. We use that bounding box to estimate where gripper jaws should be when gripping a labware, and raise errors if the actual jaw position differs too much. However, the labware bounding box is not on its own a good estimate of the gripper jaw position for two reasons: taper, and lips. First, many labwares have a little lip on the bottom; the bounding box must include it, and it's never relevant to the gripper interaction. Second, many labwares taper slightly from bottom to top, and the position of the gripper jaws depends on the height at which the gripper grips the labware.

Both of these elements have nothing to do with the wells and therefore aren't included in the gripper definition. Plus, they're a pain to measure by hand, which is what often happens with custom labware. Rather than making more accurate outer dimension data required, we can establish bounds of accuracy - a size that the labware definitely will not exceed (the labware bounding box) and a size that definitely will not exceed the labware. A good candidate for that second one is a bounding box established by the _wells_. The wells are inside the labware, so if you create a box that spans the outer edges of all the wells then you get a reasonable lower bound for the labware side.

Once we have those bounds, we can improve the gripper error calculation to take them into account. A grip is only an error if the jaw position (minus remaining uncertainty, see below) is smaller than the well bounding box or larger than the labware bounding box.

This is in ca25e3971e7569458bd268ef38d70d150003520e

## How much did the gripper jaws flex?

The gripper jaws are light, and the mechanism that drives them is light, and they apply a fair amount of force very far from where they're attached to the gripper body. That means that when the gripper is applying force through its jaws, the pads on the jaws will be farther out than the position of the motor above because the jaws will flex or rotate under the force. That difference in position can only be estimated and characterized, not sensed, and is the source of the grip error sense tolerance that remains in shared-data.

This is (implicitly) in fa1e3511e1e92505f22e3dadf0cd219686ec2fed 

## Also hard limits

Finally, though, there are some reliable hard limits. If the gripper is all the way open or all the way closed, it almost certainly is not gripping something with the required amount of force. If it's all the way open, then some motion error must have occurred or the jaws are just stuck. Even if there's a very large labware that barely fit inside the jaws, the flexing of the jaws means that the sensed position will be inboard of the stops. If the gripper is all the way closed, even if the jaws are technically around something and flexing, then in everything but the most corner of corner cases the hard stops have hit before the jaws have exerted the desired amounts of force. That means that we can confidently say that if we're at the limits, we failed. 

This is in d357035a1bef873086c39060b594c41c5d2e7649

# Testing
- Find grippers that are near the manufacturing tolerance bounds for open jaw position
- Successfully move some labware
- Get an error when taking the labware away

Closes RET-1416

sfoster1 committed:
feat(api): autocalibrate gripper jaws

The physical assembly of the gripper has a dimensionally accurate hard
stop when the gripper jaws are closed. Unfortunately, since we define
"home position" partially as "out of the way", we home the jaws open,
and zero the encoder there - the logical position of the gripper jaws
when closed is driven by distance travelled from open, and open can vary
unit-to-unit relative to closed.

The way to fix this is an internal calibration, because we can sense the
distance travelled by the jaws when moving between the stops at
open (which vary) and closed (which don't). We can then save the
difference between the expected distance and the nominal distance, and
it's a very quick thing to do - just close the gripper jaws once.

So, let's implement that as something that implicitly happens when you
home if the data isn't known. The offset won't change over time so we
only really have to do it once; we could add a whole new file to save
this in or we could just do it the first time we home every logical
gripper instance and save it in memory.

The new offset varies between 0ish and 5ish mm, always positive - the
gripper units will always open to _at least_ 92mm but maybe more.

Once we know the offset of a specific gripper, then we won't be
implicitly folding the offset into the position of the jaws and
therefore our gripper error testing, which fixes some error false
positives.

The other source of error on the gripper is the amount that the jaws
flex from the pads at the bottom to the mounts at the top under the
force of holding the labware. That value is the basis of the error
threshold in the configs.

sfoster1 committed:
feat(api): labware size bounds for gripper errors

The remaining major source of error in gripper error detection is the
size of labware. We specify the labware size when trying to detect if
the gripper has missed the labware or not, but we specify the size of
the labware using the labware's bounding box. That's often not the
actual size of the labware that's relevant because of two factors:
first, the bounding box definitionally includes anything like a lip
around the bottom of the labware that the gripper never touches; and
second, the labware may change in size - for instance, flex tipracks
taper from bottom to top.

Without adding all that data to the labware definition, one other great
touchpoint we have for labware size is where the wells are. They'll
definitionally be inside the bounding box (because they're inside the
labware) and while they may be surprisingly far inside the labware, or
surprisingly non-centered within the labware, as long as we assert in
the error checking that we will fail if the gripper gets _all the way
closed_, we can use the size of a bounding box implied by the positions
and sizes of the wells to offer a lower bound on how big the labware can
be at the point of the grip.

sfoster1 committed:
feat(api): absolute boundary errors for gripper

Notwithstanding the uncertainty in the last two commits, one thing about
the gripper is definitely true: if it is at either hard stop, open or
closed, then it hasn't gripped anything. Maybe that's because the thing
it tried to grip was just out right too narrow or too wide to work
properly, but nothing is held within the jaws.

ahiuchingau commented on api/src/opentrons/hardware_control/instruments/ot3/gripper.py:
```diff
@@ -169,6 +177,28 @@ def gripper_id(self) -> str:
     def reload_configurations(self) -> None:
         return None
 
+    def update_jaw_open_position_from_closed_position(
```
Should we sanity-check the `jaw_at_closed` value in case we're trying to save the value when the gripper jaw is stuck?

sfoster1 replied:
that is probably a good idea

sfoster1 replied:
https://github.com/Opentrons/opentrons/pull/14361/commits/7b09895902964d70fb57add749c83964924f1e45

ahiuchingau reviewed (changes requested):
This is great! I love the implementation of the well bounding box.

I think we should add a check for the `jaw_at_closed` before updating it though

sfoster1 committed:
sanity check jaw displacement

sfoster1 committed:
jaw flex is 3mm max allowable at the factory

ahiuchingau reviewed (approved):
LGTM! 😲🙌 at the 3 mm

sfoster1 committed:
refactor: move hardware logic into backends

This is pretty awful, but it's the only way to make this work with
simulation:
- we want the ot3api to be the same between simulating and not
simulating
- we want the backend to behave reasonably between simulating and not
simulating
- we want to keep some nice sanity checks

Even if we move the sanity checks down into the backend, we'd need to
tell them what's sane - the backends don't typically look at the
configuration for instruments.

This setup is a little ugly but it's really fundamentally about moving a
pure function around, and injecting some context knowledge into a
simulator as one occasionally has to do.

sfoster1 committed:
accidentally captured some patches

ryanthecoder committed:
chore(hardware-testing): update gravimetric patch files

sfoster1 committed:
test needs to handle calibration

ryanthecoder committed:
fix gripper test force QC-test

SyntaxColoring commented on api/src/opentrons/protocol_engine/execution/labware_movement.py:
```diff
@@ -156,10 +156,18 @@ async def move_labware_with_gripper(
-                                labware_id
-                            ).y
+                            expected_grip_width=labware_bbox.y,
+                            grip_width_uncertainty_wider=abs(
+                                max(well_bbox.y - labware_bbox.y, 0)
+                            ),
+                            grip_width_uncertainty_narrower=abs(
+                                min(well_bbox.y - labware_bbox.y, 0)
+                            ),
```
Is the `abs` in this `abs(max(...))` construction just for symmetry with the `min` one? As far as I can tell it's not doing anything?

sfoster1 replied:
The idea here is that all these values are _numerically_ positive, the semantic difference is in the names. So just to make sure that's always true, `abs`

SyntaxColoring commented on api/src/opentrons/protocol_engine/execution/labware_movement.py:
```diff
@@ -156,10 +156,18 @@ async def move_labware_with_gripper(
-                                labware_id
-                            ).y
+                            expected_grip_width=labware_bbox.y,
+                            grip_width_uncertainty_wider=abs(
+                                max(well_bbox.y - labware_bbox.y, 0)
+                            ),
+                            grip_width_uncertainty_narrower=abs(
+                                min(well_bbox.y - labware_bbox.y, 0)
+                            ),
```
Given that we're calling `labware_bbox.y` the expected grip width, should `grip_width_uncertainty_wider` be 0?

For example, say you have a labware that's 86 mm at its widest and whose wells form a 76 mm wide rectangle. This calculation for `grip_width_uncertainty_narrower` says that the gripper should be prepared for the labware to be as little as 76 mm wide, which makes sense. But this calculation for `grip_width_uncertainty_wider` says that the gripper should be prepared for the labware to be as much as 96 mm wide, which is unnecessarily uncertain?

sfoster1 replied:
It should be 0 and in practice it will be zero - the calculation you're calling out is the reason for the `max`. If `well_bbox.y` = 76 and `labware_bbox.y` = 86, `max(well_bbox.y - labware_bbox.y, 0) = max(76 - 86, 0) = max(-10, 0) = 0`

SyntaxColoring reviewed (commented):
A couple of questions, but the Protocol Engine parts otherwise looks good to me. Thanks!

vegano1 reviewed (approved):
This is awesome

sfoster1 committed:
more accidentally altered settings
